### PR TITLE
LPS-80310 Improvements

### PR DESCRIPTION
**Pull request by ngriffin7a**

### CODE REVIEW

**Comment by ngriffin7a** [comment]
ci:test

**Comment by ngriffin7a** [comment]
@shuyangzhou This PR simply fortifies `HeaderResponse.addDependency(...)` and makes some small fixes. To my knowledge none of our out-of-the-box portlets utilizes this feature. These failures must be unrelated.

Please review when you get an opportunity. Thanks, Neil.

**Comment by ngriffin7a** [comment]
ci:reopen

**Comment by ngriffin7a** [comment]
@shuyangzhou This PR simply fortifies `HeaderResponse.addDependency(...)` and makes some small fixes. To my knowledge none of our out-of-the-box portlets utilizes this feature. These failures must be unrelated.

Please review when you get an opportunity. Thanks, Neil.

File: portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java [hunk 1]
```diff
@@ -259,6 +292,26 @@ else if (_printWriter != null) {
 		}
 	}
 
+	private String _addClosingTags(String xml) {
```
Comment (shuyangzhou) [inline portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java]: @ngriffin7a these regex replacements are very slow, please change them to plain string manipulation and resend, thanks!

Comment (ngriffin7a) [reply portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java]: @shuyangzhou The slower regex replacements only execute if for example `if (xml.contains("<link")) {` returns true. I think the elegance of the regex is very nice, and the replacement only operates on a single line of text. Given those points, would you still like me to refactor or can we stay with the regex?

Comment (shuyangzhou) [reply portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java]: @ngriffin7a can you add a new commit to remove the regex? I will see how bad it is after, it is too complex, we can stay with regex. But I prefer no regex if we can.

**Comment by ngriffin7a** [comment]
See https://github.com/shuyangzhou/liferay-portal/pull/6597

**Pull request closed without merging**